### PR TITLE
Add test for duplicates errors codes in config files

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
## What does this PR do ?

This PR adds a test for errors config files by testing if `domains` codes are unique.
Same for subdomains, and errors.

### How should this be manually tested?

Duplicate a code either for the `domain`, `subdomain` or an `error`, then start Kuzzle.
For instance you can open the `api.json` file, and give the same code for `document` and `index` subdomains
You should have an error
Modifying these json config files doesnt trigger the hot reload
### Other changes

Finds duplicate and fix it

### CODE REVIEW

File: lib/config/error-codes/index.js [hunk 1]
```diff
@@ -85,8 +91,27 @@ function checkErrorCodes(errorCodes) {
...
+    assert(isUnique(codesSubdomains), `Error configuring file : You have a subdomain code duplicate in ${domain}.`);
+    codesSubdomains.length = 0;
   }
+  assert(isUnique(codesDomains), 'Error configuring file : You have a domain code duplicate.');
+  codesDomains.length = 0;
+}
+
+function isUnique(codesArray) {
+  const temp = [];
+  let it;
```
Comment (kuzzle) [inline lib/config/error-codes/index.js]: MAJOR Move the declaration of "it" to line 108. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2392)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: lib/config/error-codes/index.js [hunk 2]
```diff
@@ -85,8 +91,27 @@ function checkErrorCodes(errorCodes) {
...
+    codesSubdomains.length = 0;
   }
+  assert(isUnique(codesDomains), 'Error configuring file : You have a domain code duplicate.');
+  codesDomains.length = 0;
+}
+
+function isUnique(codesArray) {
+  const temp = [];
+
+  for (let it = 0; it < codesArray.length; it += 1) {
```
Comment (Aschen) [inline lib/config/error-codes/index.js]: Use `for (const ... of ...)`

File: lib/config/error-codes/index.js [hunk 3]
```diff
@@ -85,8 +91,27 @@ function checkErrorCodes(errorCodes) {
...
+      codesErrors.length = 0;
     }
+    assert(isUnique(codesSubdomains), `Error configuring file : You have a subdomain code duplicate in ${domain}.`);
+    codesSubdomains.length = 0;
   }
+  assert(isUnique(codesDomains), 'Error configuring file : You have a domain code duplicate.');
+  codesDomains.length = 0;
+}
+
+function isUnique(codesArray) {
```
Comment (alexandrebouthinon) [inline lib/config/error-codes/index.js]:
```suggestion
function isUnique(arr) {
  return arr.filter((itm, i) => return arr.lastIndexOf(itm) === i && arr.indexOf(itm) !== i).length === 0;
}
```

Comment (alexandrebouthinon) [reply lib/config/error-codes/index.js]: Smaller implementation :slightly_smiling_face:

Comment (Yoann-Abbes) [reply lib/config/error-codes/index.js]: I did it another way with lodash

Comment (scottinet) [inline lib/config/error-codes/index.js]: I find your implementation very complicated. 

1. You can simply use Sets instead of Arrays and in your code, before pushing code values in it, assert that it doesn't already exist
2. Use scopes: don't declare a global container just to reinitialize it each time a loop iterates, instead let JS destroy containers when exiting a scope and create a fresh new one every time you need one
3. Make the error messages show what are the duplicates: nothing is more irritating than "oh, there is an error somewhere in that file, good luck finding it" ;-)
4. You named the name of a domain `domain`, and the domain object `domainName`. This should be permutated (same remark applies to subdomains and errors).

All in all, that should give you something like that:

```js
function checkErrorCodes(errorCodesFiles) {
  const domainCodes = new Set();

  for (const domainName of Object.keys(errorCodesFiles)) {
    const domain = errorCodesFiles[domainName];

    assert(
      !domainCodes.has(domain.code), 
      `Error configuration file : domain code ${domain.code} is not unique`);

    domainCodes.add(domain.code);

    // other asserts

    const subdomainCodes = new Set();

    for (const subdomainName of Object.keys(domain.subdomains)) {
      const subdomain = domain.subdomains[subdomainName];

      assert(
        !subdomainCodes.has(subdomain.code), 
        `Error codes file : subdomain code ${subdomain.code} is not unique (domain: ${domain.code})`);

      subdomainCodes.add(subdomain.code);

      // other asserts

      const errorCodes = new Set();

      for (const errorName of Object.keys(subdomain.errors)) {
        const error = subdomain.errors[errorName];

        assert(
          !errorCodes.has(error.code), 
          `Error codes file : error code ${error.code} is not unique (domain: ${domain.code}, subdomain: ${subdomain.code})`);

        errorCodes.add(error.code);

        // other asserts
      }
    }
  }
}
```

And that way, you can remove the isUnique method.

One last remark: you should probably separate domain, subdomain and error checks in distinct functions. 
This `checkErrorCodes` is starting to be too large.

Comment (Yoann-Abbes) [reply lib/config/error-codes/index.js]: I completely changed it right now @scottinet now

Comment (scottinet) [reply lib/config/error-codes/index.js]: If you separate the domain/subdomain/error checks in distinct functions, that gives something like that:

```js
function checkErrors(subdomain) {
  const errorCodes = new Set();

  for (const errorName of Object.keys(subdomain.errors)) {
    const error = subdomain.errors[errorName];

    assert(
      !errorCodes.has(error.code), 
      `Error codes file : error code ${error.code} is not unique (domain: ${domain.code}, subdomain: ${subdomain.code})`);

    errorCodes.add(error.code);

    // other asserts
  }
}

function checkSubdomains(domain) {
  const subdomainCodes = new Set();

  for (const subdomainName of Object.keys(domain.subdomains)) {
    const subdomain = domain.subdomains[subdomainName];

    assert(
      !subdomainCodes.has(subdomain.code), 
      `Error codes file : subdomain code ${subdomain.code} is not unique (domain: ${domain.code})`);

    subdomainCodes.add(subdomain.code);

    // other asserts

    checkErrors(subdomain);
  }  
}

function checkDomains(errorCodesFiles) {
  const domainCodes = new Set();

  for (const domainName of Object.keys(errorCodesFiles)) {
    const domain = errorCodesFiles[domainName];

    assert(
      !domainCodes.has(domain.code), 
      `Error configuration file : domain code ${domain.code} is not unique`);

    domainCodes.add(domain.code);

    // other asserts

    checkSubdomains(domain);
  }
}
```

And I find that far more readable.

Comment (alexandrebouthinon) [reply lib/config/error-codes/index.js]: Why do not use this 
```

Comment (scottinet) [reply lib/config/error-codes/index.js]: Why do you persist to emulate a Set with an Array?

Comment (alexandrebouthinon) [reply lib/config/error-codes/index.js]: I like you suggestion but I do not understand why you said I'm emulating Set with Array? :disappointed_relieved:

Comment (scottinet) [reply lib/config/error-codes/index.js]: Yoann's code pushes value to arrays, and at the end of each loop, it checks for uniqueness in each one of those arrays.
With Set, you can do the same thing, with less code, and all you have to do is to check before adding a new value to the set.

So, since both approaches seem strictly equivalent (to me at least), then I consider that using array for that purpose is like emulating sets.

Comment (alexandrebouthinon) [reply lib/config/error-codes/index.js]: Thanks for explanations :slightly_smiling_face:

Comment (Yoann-Abbes) [reply lib/config/error-codes/index.js]: Thanks :O
